sandbox/apparmor/notify/listener/listener.go

```diff
@@ -0,0 +1,473 @@
+			poll.Close()
+		}
+	}()
+	if err = poll.Register(int(notifyFile.Fd()), epoll.Readable); err != nil {
+		return nil, fmt.Errorf("cannot register epoll on %q: %v", path, err)
+	}
+
+	listener = &Listener{
+		reqs: make(chan *Request, 1),
+
+		notifyFile: notifyFile,
+		poll:       poll,
+	}
+	return listener, nil
+}
+
+// Close stops the listener and closes the kernel communication file.
+// Returns once all waiting goroutines terminate and the communication file
+// is closed.
+func (l *Listener) Close() error {
+	origStatus := atomic.SwapUint32(&l.status, statusClosed)
+	switch origStatus {
+	case statusReady:
+		// A goroutine was never spawned with the listener tomb.
+		// Spawn one, so the tomb can die.
+		l.tomb.Go(func() error {
+			<-l.tomb.Dying()
+			return nil
+		})
+	case statusRunning:
+	case statusClosed:
+		l.tomb.Wait()
+		return ErrAlreadyClosed
+	}
+	l.tomb.Kill(ErrClosed)
+	// Close epoll instance to stop the run loop waiting on epoll events
+	err1 := l.poll.Close()
+	// Wait for all waiters to send back deny responses to the kernel
+	l.tomb.Wait()
+	err2 := l.notifyFile.Close()
+	if err1 != nil {
+		return err1
+	}
+	return err2
+}
+
+// Reqs returns a read-only channel through which requests may be received.
+// The channel is closed when the Close() method is called or an error occurs.
+func (l *Listener) Reqs() <-chan *Request {
+	return l.reqs
+}
+
+// Run reads and dispatches kernel requests until the listener is closed.
+//
+// Run should only be called once per listener object. If called more than once,
+// Run returns an error. Otherwise, waits until the listener stops, and returns
+// the cause as an error. If the listener was intentionally stopped via the
+// Close() method, returns nil.
+func (l *Listener) Run() error {
+	if !atomic.CompareAndSwapUint32(&l.status, statusReady, statusRunning) {
+		currStatus := atomic.LoadUint32(&l.status)
+		switch currStatus {
+		case statusRunning:
+			return ErrAlreadyRun
+		case statusClosed:
+			return ErrAlreadyClosed
+		default:
+			return fmt.Errorf("listener has unexpected status: %d", currStatus)
+		}
+	}
+	// This is the first and only time calling Run().
+	// Even if Close() kills the tomb before l.tomb.Go() occurs, a panic will
+	// not occur, since this new goroutine will (immediately after l.tomb.Err()
+	// is called) return and close the tomb's dead channel, as it is the last
+	// and only tracked goroutine to terminate.
+	l.tomb.Go(func() error {
+		var err error
+		for {
+			err = l.tomb.Err()
+			if err != tomb.ErrStillAlive {
+				break
+			}
+			err = l.runOnce()
+			if err != nil {
+				break
+			}
+		}
+		return err
+	})
+	// Wait for an error to occur or the listener to be explicitly closed.
+	<-l.tomb.Dying()
+	// Close the listener, in case an internal error occurred.
+	// Close() waits until all waiters have sent back deny responses to the
+	// kernel, so it is safe to call without first calling l.tomb.Wait().
+	l.Close()
+	close(l.reqs)
```

Comment:
probably a comment should say that by this point nothing will send over l.reqs so it is safe to close